File: src/lib/ai/manager.ts
import { createServerFn } from "@tanstack/react-start";
import { MANAGER_SYSTEM_PROMPT } from "@/lib/ai/prompt";
import type { ManagerReply, Quote } from "@/lib/ai/types";

type ChatTurn = { role: "user" | "assistant"; content: string };

type Payload = {
  messages: ChatTurn[];
  context?: string;
};

const OPENROUTER_URL = "https://openrouter.ai/api/v1/chat/completions";
const OPENROUTER_MODEL = "qwen/qwen3.7-flash";
const XAI_URL = "https://api.x.ai/v1/chat/completions";
const XAI_MODEL = "grok-4.5";

function isQuote(value: unknown): value is Quote {
  if (!value || typeof value !== "object") return false;
  const q = value as Record<string, unknown>;
  if (typeof q.title !== "string" || typeof q.timeline !== "string") return false;
  if (typeof q.total !== "number" || !Array.isArray(q.items)) return false;
  return q.items.every((item) => {
    if (!item || typeof item !== "object") return false;
    const row = item as Record<string, unknown>;
    return typeof row.name === "string" && typeof row.price === "number";
  });
}

function extractJsonObject(text: string): unknown {
  const fenced = text.match(/```(?:json)?\s*([\s\S]*?)```/);
  const raw = (fenced?.[1] ?? text).trim();
  const start = raw.indexOf("{");
  const end = raw.lastIndexOf("}");
  if (start < 0 || end <= start) throw new Error("no json");
  return JSON.parse(raw.slice(start, end + 1)) as unknown;
}

function parseManager(text: string): { message: string; questions: string[]; quote: Quote | null } {
  try {
    const data = extractJsonObject(text) as Record<string, unknown>;
    const message =
      typeof data.message === "string" && data.message.trim()
        ? data.message.trim()
        : text.trim();
    const questions = Array.isArray(data.questions)
      ? data.questions.filter((q): q is string => typeof q === "string" && q.trim().length > 0)
      : [];
    const quote = isQuote(data.quote) ? data.quote : null;
    if (quote) {
      const sum = quote.items.reduce((n, i) => n + i.price, 0);
      if (Math.abs(sum - quote.total) > 1) quote.total = sum;
    }
    return { message, questions, quote };
  } catch {
    return { message: text.trim(), questions: [], quote: null };
  }
}

type CompatBody = {
  model: string;
  messages: { role: "system" | "user" | "assistant"; content: string }[];
  temperature: number;
  max_tokens: number;
};

async function postJson(opts: {
  url: string;
  apiKey: string;
  body: CompatBody;
  extraHeaders?: Record<string, string>;
  socks?: string;
}): Promise<string> {
  const payload = JSON.stringify(opts.body);
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${opts.apiKey}`,
    ...opts.extraHeaders,
  };

  if (opts.socks) {
    const { execFile } = await import("node:child_process");
    const { promisify } = await import("node:util");
    const execFileAsync = promisify(execFile);
    const proxy = opts.socks.replace(/^socks5h?:\/\//i, "");
    const headerArgs = Object.entries(headers).flatMap(([k, v]) => ["-H", `${k}: ${v}`]);
    const { stdout } = await execFileAsync(
      "curl",
      [
        "-sS",
        "-g",
        "--max-time",
        "90",
        "--socks5-hostname",
        proxy,
        "-X",
        "POST",
        ...headerArgs,
        "-d",
        payload,
        opts.url,
      ],
      { maxBuffer: 2_000_000 },
    );
    return stdout;
  }

  const res = await fetch(opts.url, {
    method: "POST",
    headers,
    body: payload,
  });
  const raw = await res.text();
  if (!res.ok) {
    throw new Error(`API ${res.status}: ${raw.slice(0, 280)}`);
  }
  return raw;
}

function readChoice(raw: string): string {
  const body = JSON.parse(raw) as {
    error?: { message?: string };
    choices?: { message?: { content?: string } }[];
  };
  if (body.error?.message) throw new Error(body.error.message);
  const text = body.choices?.[0]?.message?.content ?? "";
  if (!text.trim()) throw new Error("Пустой ответ модели");
  return text;
}

async function complete(messages: CompatBody["messages"]): Promise<string> {
  const openrouterKey = process.env.OPENROUTER_API_KEY;
  const xaiKey = process.env.XAI_API_KEY;
  const socks = process.env.TOR_SOCKS_PROXY || process.env.SOCKS_PROXY;
  const body: CompatBody = {
    model: OPENROUTER_MODEL,
    messages,
    temperature: 0.4,
    max_tokens: 1400,
  };

  if (openrouterKey) {
    try {
      const raw = await postJson({
        url: OPENROUTER_URL,
        apiKey: openrouterKey,
        body,
        socks,
        extraHeaders: {
          "HTTP-Referer": "https://codwey.su",
          "X-Title": "Codwey",
        },
      });
      return readChoice(raw);
    } catch (err) {
      if (!xaiKey) throw err;
    }
  }

  if (xaiKey) {
    const raw = await postJson({
      url: XAI_URL,
      apiKey: xaiKey,
      body: { ...body, model: XAI_MODEL },
    });
    return readChoice(raw);
  }

  throw new Error("AI is not available in this environment");
}

export const sendManagerMessage = createServerFn({ method: "POST" })
  .validator((input: Payload) => input)
  .handler(async ({ data }): Promise<ManagerReply> => {
    const trimmed = data.messages
      .filter((m) => m.content.trim())
      .slice(-16)
      .map((m) => ({ role: m.role, content: m.content.slice(0, 4000) }));

    if (trimmed.length === 0) {
      return { ok: false, error: "Напишите сообщение" };
    }

    const lastUser = [...trimmed].reverse().find((m) => m.role === "user");
    if (!lastUser) return { ok: false, error: "Напишите сообщение" };

    const system = data.context
      ? `${MANAGER_SYSTEM_PROMPT}\n\nКонтекст страницы:\n${data.context.slice(0, 2500)}`
      : MANAGER_SYSTEM_PROMPT;

    try {
      const text = await complete([
        { role: "system", content: system },
        ...trimmed,
      ]);
      const parsed = parseManager(text);
      return { ok: true, ...parsed };
    } catch (err) {
      const message = err instanceof Error ? err.message : "Не удалось ответить";
      return { ok: false, error: message };
    }
  });

File: vds/manager.py
"""ИИ-менеджер Вей: OpenRouter через Tor. Ключ только в .env на VDS."""
from __future__ import annotations

import json
import re
from typing import Any

import httpx
from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.config import settings

router = APIRouter()

SYSTEM_PROMPT = """Ты — Вей, ИИ-менеджер студии Codwey (codway.su).
Говоришь по-русски, коротко, по делу, без канцелярита и без эмодзи. На «ты».

Codwey продаёт готовые сайты, Telegram-боты, веб-приложения, Roblox-игры и делает то же на заказ.
Цены ниже Авито: готовые сборки и небольшие кастомные работы.

Мы делаем только сайт, бота, приложение или игру. Хостинг, домен, сервер, токены и всё остальное лежит на клиенте.

Каталог готового (₽): PulseShop 1490, TicketGate 1290, QuizPulse 990, SitePulse 2490, FolioKit 1990, ShopPeek 3490, PWA Kit 3990, ObbyStart 1290, TycoonLite 2490.

Правила цен на заказ:
- Telegram-бот: база 990, магазин +400–800, оплата +400, админка +300.
- Лендинг 1990–3490. Магазин-витрина 3490–5990.
- PWA от 3990. Roblox обби от 1290. Парсеры от 790.
- Не завышай. Валюта только ₽. Не обещай App Store / Google Play.

Если бриф дырявый — доспроси 1–2 вопроса. Когда данных хватает — верни quote.

ФОРМАТ — строго JSON без markdown:
{
  "message": "текст клиенту",
  "questions": ["вопрос"],
  "quote": null или {
    "title": "заголовок",
    "items": [{"name": "блок", "detail": "зачем", "price": 1490}],
    "total": 1490,
    "timeline": "5–7 дней",
    "notes": "что не входит"
  }
}
"""


class Turn(BaseModel):
    role: str
    content: str


class Payload(BaseModel):
    messages: list[Turn] = Field(default_factory=list)
    context: str | None = None


def _extract_json(text: str) -> dict[str, Any] | None:
    fenced = re.search(r"```(?:json)?\s*([\s\S]*?)```", text)
    raw = (fenced.group(1) if fenced else text).strip()
    start = raw.find("{")
    end = raw.rfind("}")
    if start < 0 or end <= start:
        return None
    try:
        data = json.loads(raw[start : end + 1])
    except json.JSONDecodeError:
        return None
    return data if isinstance(data, dict) else None


def parse_manager(text: str) -> dict[str, Any]:
    data = _extract_json(text)
    if not data:
        return {"ok": True, "message": text.strip(), "questions": [], "quote": None}
    message = data.get("message")
    message = message.strip() if isinstance(message, str) and message.strip() else text.strip()
    questions = [q for q in data.get("questions", []) if isinstance(q, str) and q.strip()]
    quote = data.get("quote")
    if not isinstance(quote, dict):
        quote = None
    else:
        items = quote.get("items")
        if not isinstance(items, list) or not items:
            quote = None
        else:
            clean = []
            for row in items:
                if not isinstance(row, dict):
                    continue
                name = row.get("name")
                price = row.get("price")
                if not isinstance(name, str) or not isinstance(price, (int, float)):
                    continue
                item = {"name": name, "price": int(price)}
                if isinstance(row.get("detail"), str):
                    item["detail"] = row["detail"]
                clean.append(item)
            if not clean:
                quote = None
            else:
                total = sum(i["price"] for i in clean)
                quote = {
                    "title": quote.get("title") if isinstance(quote.get("title"), str) else "Смета",
                    "items": clean,
                    "total": total,
                    "timeline": quote.get("timeline") if isinstance(quote.get("timeline"), str) else "по согласованию",
                    "notes": quote.get("notes") if isinstance(quote.get("notes"), str) else None,
                }
    return {"ok": True, "message": message, "questions": questions, "quote": quote}


async def complete(messages: list[dict[str, str]]) -> str:
    key = settings.openrouter_api_key
    if not key:
        raise RuntimeError("OPENROUTER_API_KEY missing")
    model = settings.openrouter_model or "qwen/qwen3.7-flash"
    async with httpx.AsyncClient(
        timeout=httpx.Timeout(90.0, connect=15.0),
        proxy="socks5://127.0.0.1:9050",
    ) as client:
        response = await client.post(
            "https://openrouter.ai/api/v1/chat/completions",
            headers={
                "Authorization": f"Bearer {key}",
                "HTTP-Referer": "https://codway.su",
                "X-Title": "Codwey",
                "Content-Type": "application/json",
            },
            json={
                "model": model,
                "messages": messages,
                "temperature": 0.4,
                "max_tokens": 1400,
            },
        )
        response.raise_for_status()
        data = response.json()
        text = data.get("choices", [{}])[0].get("message", {}).get("content") or ""
        if not str(text).strip():
            raise RuntimeError("empty model")
        return str(text)


@router.post("")
async def manager(payload: Payload) -> dict[str, Any]:
    trimmed = [
        {"role": m.role, "content": m.content[:4000]}
        for m in payload.messages
        if m.role in ("user", "assistant") and m.content.strip()
    ][-16:]
    if not trimmed or not any(m["role"] == "user" for m in trimmed):
        return {"ok": False, "error": "Напишите сообщение"}

    system = SYSTEM_PROMPT
    if payload.context:
        system += "\n\nКонтекст страницы:\n" + payload.context[:2500]

    try:
        text = await complete([{"role": "system", "content": system}, *trimmed])
        return parse_manager(text)
    except Exception:
        return {"ok": False, "error": "Не удалось ответить"}
